lgc/patch/Gfx6ConfigBuilder.cpp

```diff
@@ -1023,11 +1023,16 @@ void ConfigBuilder::buildPsRegConfig(ShaderStage shaderStage, T *pConfig) {
       static const unsigned PassThroughMode = (1 << 5);
       spiPsInputCntl.bits.FLAT_SHADE = true;
       spiPsInputCntl.bits.OFFSET |= PassThroughMode;
-    } else {
-      if (interpInfoElem.is16bit) {
+    } else if (interpInfoElem.is16bit) {
+      if (!m_pipelineState->isPackInOut()) {
         // NOTE: Enable 16-bit interpolation mode for non-passthrough mode. Attribute 0 is always valid.
         spiPsInputCntl.bits.FP16_INTERP_MODE__VI = true;
         spiPsInputCntl.bits.ATTR0_VALID__VI = true;
+      } else if (!interpInfoElem.flat) {
+        // Attribute 0 and Attribute 1 are valid when two halfs are packed in a 32-bit attribute
+        spiPsInputCntl.bits.FP16_INTERP_MODE__VI = true;
+        spiPsInputCntl.bits.ATTR0_VALID__VI = true;
+        spiPsInputCntl.bits.ATTR1_VALID__VI = true;
```

Comment:
A couple of points of concern with this hunk:
* We're setting ATTR1_VALID even when it might not be, e.g. if you only have 3 x f16 interpolants, then the last location doesn't have a valid high half and should be ATTR1_VALID = 0.
* I don't understand the logic for "flat" here: if we should keep FP16_INTERP_MODE = 0 for flat, then that should be both in the packed and non-packed mode.

More generally, I think this code should no longer have to care about isPackInOut() -- all information should be implicit in the interpInfoElem. I think what we want is to extend `FsInterpInfo` with attr0Valid and attr1Valid fields. `patchFsGenericInputImport` would set those fields based on `highHalf`.

Reply:
I previously tried to externd FsInterpInfo with a highhalf field but I came accross some failure tests. I will try it again and retest.
Pack mode needs FP16_INTERP_MODE = 0 for flat to make tests pass. I will verify your suggestion of keeping FP16_INTERP_MODE = 0 for flat in both mode.

Reply:
It seems work well as in your way. I found the previous failure is due to the wrong logic of setting attr1Valid based on highHalf for each fs input.